src/os_darwin/os_futex.c

```diff
@@ -0,0 +1,83 @@
+/*-
+ * Copyright (c) 2024-present MongoDB, Inc.
+ *	All rights reserved.
+ *
+ * See the file LICENSE for redistribution information.
+ */
+
+#include "wt_internal.h"
+#include <sys/errno.h>
+
+#define PRIVATE 1
+#include <ulock.h>
+#undef PRIVATE
+
+/*
+ * __wt_futex_wait --
+ *     Wait on the futex. The timeout is in microseconds and MUST be greater than zero.
+ */
+int
+__wt_futex_wait(
+  volatile WT_FUTEX_WORD *addr, WT_FUTEX_WORD expected, time_t usec, WT_FUTEX_WORD *wake_valp)
+{
+    WT_DECL_RET;
+    uint64_t nsec;
+
+    /* Check for overflow? */
+    nsec = (uint64_t)usec * 1000;
+
+    ret = __ulock_wait2(UL_COMPARE_AND_WAIT_SHARED | ULF_NO_ERRNO, (void *)addr, expected, nsec, 0);
```

Comment:
I think a comment here would be useful - I'm not familiar with macOS flags the same way I am with Linux, and there seems to be less online about them.

On that note - why `ULF_NO_ERRNO`? Seems like omitting it would remove the need for the "invert ret and set errno ourselves" dance.

Reply:
Originally I thought I was going to have to deal with far more errno values and was using a switch statement. I don't think the "dance" is that problematic, either way I have to clobber `errno` either in the else or for `EFAULT`. So I think it is a wash and I'm inclined to leave it but not hard on it.

Reply:
I'm OK with leaving it if there's a comment about why it's that way and the tradeoffs involved. Due to there being limited documentation (at least that I could find) it feels a little "magical" which I'd like to avoid.

Reply:
changed code flow and updated docs

Reply:
Thanks - this is easier for me to understand now.